sdk/include/opentelemetry/sdk/trace/batch_span_processor.h

```diff
@@ -0,0 +1,158 @@
+   */
+  void Shutdown(std::chrono::microseconds timeout = std::chrono::microseconds(0)) noexcept override;
+
+  /**
+   * Class destructor which invokes the Shutdown() method. The Shutdown() method is supposed to be
+   * invoked when the Tracer is shutdown (as per other languages), but the C++ Tracer only takes
+   * shared ownership of the processor, and thus doesn't call Shutdown (as the processor might be
+   * shared with other Tracers).
+   */
+  ~BatchSpanProcessor();
+
+private:
+  /**
+   * The background routine performed by the worker thread.
+   */
+  void DoBackgroundWork();
+
+  /**
+   * Exports all ended spans to the configured exporter.
+   *
+   * @param buffer - The buffer with ended spans to export
+   * @param was_force_flush_called - A flag to check if the current export is the result
+   *                                 of a call to ForceFlush method. If true, then we have to
+   *                                 notify the main thread to wake it up in the ForceFlush
+   *                                 method.
+   */
+  void Export(std::unique_ptr<common::CircularBuffer<Recordable>> &buffer,
+              const bool was_for_flush_called);
+
+  /**
+   * Called when Shutdown() is invoked. Completely drains the queue of all its ended spans and
+   * passes them to the exporter.
+   */
+  void DrainQueue();
+
+  /**
+   * Consumes and copies the appropriate amount of spans from the buffer_
+   * to a copy buffer which is then exported. This helps unblock all producers
+   * and increases the overall synchronization performance.
+   *
+   * @param was_force_flush_called - A flag to check if the current export is the result
+   *                                 of a call to ForceFlush method. If true, we consume
+   *                                 and copy the entire buffer_ to the copy buffer. Otherwise,
+   *                                 we calculate the appropriate size to export.
+   * @return A unique pointer to the copy buffer
+   */
+  std::unique_ptr<common::CircularBuffer<Recordable>> CopySpans(const bool was_force_flush_called);
+
+  /* The configured backend exporter */
+  std::unique_ptr<SpanExporter> exporter_;
+
+  /* The background worker thread */
+  std::unique_ptr<std::thread> worker_thread_;
+
+  /* Configurable parameters as per the official specs */
+  const std::chrono::milliseconds schedule_delay_millis_;
+  const size_t max_queue_size_;
+  const size_t max_export_batch_size_;
+
+  /* Synchronization primitives */
+  std::condition_variable cv_, force_flush_cv_;
+  std::mutex cv_m_, force_flush_cv_m_;
+
+  /* The buffer/queue to which the ended spans are added */
+  std::unique_ptr<common::CircularBuffer<Recordable>> buffer_;
+
+  /* Important boolean flags to handle the workflow of the processor */
+  volatile bool is_shutdown_{false};
+  volatile bool is_force_flush_{false};
+  volatile bool is_force_flush_notified_{false};
```

Comment:
This is to prevent over-optimization on certain compilers. [Discussed](https://github.com/open-telemetry/opentelemetry-cpp/pull/195#discussion_r460344024) with @reyang

Reply:
`std::atomic_flag` is better suited for this use case. The use of `volatile` in use cases like this is controversial (see [this discussion](https://stackoverflow.com/questions/4557979/when-to-use-volatile-with-multi-threading)).

Reply:
Alright. I had actually planned on using `std::atomic<bool>` after reading that exact same [discussion](https://stackoverflow.com/questions/4557979/when-to-use-volatile-with-multi-threading), and had earlier suggested it in [my discussion with reiley](https://github.com/open-telemetry/opentelemetry-cpp/pull/195#discussion_r460344024), but I deleted that comment because of delay in feedback and instead just changed it to volatile for the time being.

I was wondering if `std::atomic<bool>` was the better choice here since there is no way to just retrieve the value of `std::atomic_flag` without modifying it. (speaking with reference to the `std::atomic_flag::test_and_set()` function)